### DISCUSSION TITLE
Installation in windows for dummies

### DISCUSSION THREAD

**Discussion by srcfernandes** [jonescompneurolab/hnn-core#628] [Installation] [answered]
Dear HNN team

I am interested in local neural networks modelling.
I am familiar with other types of programming, such as matlab and just starting with python.

My main problem is to know how to make installations with anaconda, how are the steps procedures, prompts to open, paths to add and how.... all this kind of stuft. I tried to install hnn-core with a help of a friend, but it didn't work. 

I kindly ask you if you could provide me a step-by-step instructions on how to install HNN core in windows, test and launch hnn-core (a kind of "installation for dummies"). If this is already in here, where should I look? 

Thank you so much!

Sofia

**Comment by jasmainak** [comment]
@srcfernandes can you be a bit more explicit about "it didn't work". What/where did you fail?

Did you follow the steps here: https://jonescompneurolab.github.io/hnn-core/stable/index.html ?

**Comment by srcfernandes** [comment]
Dear Mainak Jas,


I installed anaconda3 (local installation).

Then, I installed neuron as recommended using the precompiled installer for windows 10.

I selected the anaconda3 dir as installation dir for neuron (is this correct?)


Then, I opened the Anaconda Prompt and checked, as you recommended:


(base) C:\Users\FML>python -c 'import neuron;'
  File "<string>", line 1
    'import
    ^
SyntaxError: unterminated string literal (detected at line 1)

(base) C:\Users\FML>

Since it was a syntax error, I tried

(base) C:\Users\FML>python -c "import neuron"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'neuron'

I did not move forward, since I was not able to check neuron installation.
I am waiting for confirmation of acceptance in the Neuron forum to ask them.
However, if you could help me with is, it would be great. Maybe I am missing some steps.

Many thanks, best,

Sofia



________________________________

**Comment by jasmainak** [comment]
I don't think you need to set anaconda3 as the install directory for Neuron. @rythorpe do you have any advice? 

I hope they accept you in the forum soon, otherwise we can tag one of the Neuron developers here ...

**Comment by rythorpe** [comment]
@srcfernandes You should be able to install NEURON anywhere on your machine and it work with anaconda, but I'd recommend installing NEURON at root or wherever it tries to install by default.

**Comment by srcfernandes** [comment]
Dear Ryan


I re-installed Neuron at root (the defaul) as recommended.


When I open the Anaconda prompt and do, as recommended in your installation page (https://jonescompneurolab.github.io/hnn-core/stable/index.html#installation)


python -c 'import neuron;'
  File "<string>", line 1
    'import
    ^
SyntaxError: unterminated string literal (detected at line 1)


But if I replace ' for ", it does not result in errors


(base) C:\Users\FML>python -c "import neuron;"

(base) C:\Users\FML>

Nothing happens, is this OK?

I then installed hnn_core[gui]:

(base) C:\Users\FML>pip install hnn_core[gui]

then, I tested and no errors were reported

(base) C:\Users\FML>python -c "import hnn_core"

(base) C:\Users\FML>

but when I try to start the gui according to the instruction provided in the webpage, I get an error

(base) C:\Users\FML>hnn-gui

'hnn-gui' is not recognized as an internal or external command,
operable program or batch file.

I tried other things, with the same error

(base) C:\Users\FML>hnn_core
'hnn_core' is not recognized as an internal or external command,
operable program or batch file.

(base) C:\Users\FML>hnn
'hnn' is not recognized as an internal or external command,
operable program or batch file.

(base) C:\Users\FML>hnn_core[gui]
'hnn_core[gui]' is not recognized as an internal or external command,
operable program or batch file.


What have I missed here? I am doing this all in the Anaconda prompt.
Do I need do do something like "conda activate"?
I also send a text with the outputs that I get during hnn-core installation.

Also, I would like to use Jupyter lab to run the scripts.
What should be the steps for starting this?

Thank you for all your help, best

Sofia





________________________________

**Comment by srcfernandes** [comment]
Dear Ryan,


Meanwhile, I follow some indications I get from Sara Piersigilli (she already contacted you) and did


pip install --upgrade https://api.github.com/repos/jonescompneurolab/hnn-core/zipball/master


It runned fine.


I then did


(base) C:\Users\FML>python -c "import hnn_core"

and got nothing (so it must be OK)


and did


(base) C:\Users\FML>hnn-gui


Here, it started the hnn_widget but it was empthy. There were several errors in the prompt, I send all the outputs in attachment.


Can you help me?


Many thanks, best,


Sofia








________________________________

**Comment by rythorpe** [comment]
A few things:

1. Using single quotes vs double quotes is different depending on if you're running from a Windows cmd shell, so I think you're right that at times you need to use double quotation marks.
2. In order to run the `hnn-gui` command, try calling `start hnn-gui`.
3. `hnn_core` is a python module, not an executable, so you won't be able to call it directly outside of a python interpreter.

**Comment by rythorpe** [comment]
Oh, and it looks like your `voila` and/or `ipywidgets` modules might be the wrong versions (see our [README](https://jonescompneurolab.github.io/hnn-core/stable/index.html)). Try downgrading to the specified versions or starting a new conda environment.

**Comment by srcfernandes** [comment]
Dear Ryan


Can you help me on how do I proceed to make the procedures you are indicating?


1) downgrading to the specific versions?

2) starting a new conda environment?


I am new to anaconda and conda, so I do not know how to do this.


Should the problem be because I did this first



pip install hnn_core[gui]


and then this


pip install --upgrade https://api.github.com/repos/jonescompneurolab/hnn-core/zipball/master


Thank  you for all your help and attention.


Best,


Sofia



________________________________

**Comment by jasmainak** [comment]
@srcfernandes take a look at the conda documentation:

https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#creating-an-environment-with-commands

Something like:

```sh
$ conda create -n hnn_gui python=3.9
```

should do the trick. Followed by:

```sh
$ conda activate hnn_gui
```

I don't think you need to separately downgrade any package since the `hnn_core` install should automatically install the right versions of these packages using the constraints specified in `setup.py`. Just try redoing the install in a fresh environment and check if it works ...

**Comment by jasmainak** [comment]
Btw, it would be helpful if you could format the code blocks using markdown when asking for help. So it is more readable:

https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code

Also googling your warnings led me to these:

https://github.com/voila-dashboards/voila/issues/1271
https://github.com/jonescompneurolab/hnn-core/issues/585

Perhaps you can find something useful there?

**Comment by srcfernandes** [comment]
Dear Mainak Jas


Thank you for your help. I apologize for the following question, I am really new on all these instalation procedures.


I did  the conda commands:

$ conda create -n hnn_gui python=3.9

$ conda activate hnn_gui

but when I try to run the hnn-gui in the default dir, I get the following error:

(I used different combinations)


(hnn_gui) C:\Users\FML>hnn-gui
'hnn-gui' is not recognized as an internal or external command,
operable program or batch file.

(hnn_gui) C:\Users\FML>hnn-core[gui]
'hnn-core[gui]' is not recognized as an internal or external command,
operable program or batch file.

(hnn_gui) C:\Users\FML>hnn_core[gui]
'hnn_core[gui]' is not recognized as an internal or external command,
operable program or batch file.

(hnn_gui) C:\Users\FML>hnn_gui
'hnn_gui' is not recognized as an internal or external command,
operable program or batch file.


Even if I try the same inside hnn_gui directory, I get the same error.


I noticed that there are 2 executable files inside hnn_gui directory, "venvlauncher.exe" and "venvwlauncher.exe".

When I type these, I get


>> No pyvenv.cfg file


Do I need to unistall hnn-core and install it again inside this new environment?

This does not apply to neuron, right?


I apologize for these many questions.


Thank you so much for your help.


Best,


Sofia








________________________________

**Comment by srcfernandes** [comment]
Dear Mainak Jas


Meanwhile, I tried the following:


(hnn_gui) C:\Users\FML>pip uninstall hnn_core[gui]


(hnn_gui) C:\Users\FML>pip install --upgrade https://api.github.com/repos/jonescompneurolab/hnn-core/zipball/master


All went fine.


When I type "hnn-gui", I get the following error


(hnn_gui) C:\Users\FML>hnn-gui
Traceback (most recent call last):
  File "C:\Users\FML\anaconda3\envs\hnn_gui\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\FML\anaconda3\envs\hnn_gui\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\FML\anaconda3\envs\hnn_gui\Scripts\hnn-gui.exe\__main__.py", line 4, in <module>
  File "C:\Users\FML\anaconda3\envs\hnn_gui\lib\site-packages\hnn_core\gui\__init__.py", line 1, in <module>
    from .gui import HNNGUI, launch
  File "C:\Users\FML\anaconda3\envs\hnn_gui\lib\site-packages\hnn_core\gui\gui.py", line 14, in <module>
    from IPython.display import IFrame, display
ModuleNotFoundError: No module named 'IPython'


Should this be a problem with PATH (to get IPython working)?


Many, many thanks for all your help, best


Sofia





________________________________

**Comment by srcfernandes** [comment]
Dear Mainak Jas


I am going step by step and trying to find what is missing.


I opened the anaconda navigator and installed IPython packages on hnn_gui environment.

I am not sure that this is the procedure, I hope so.


When I did "hnn-gui", it seems that only voila is missing:


(hnn_gui) C:\Users\FML>hnn-gui
Traceback (most recent call last):
  File "C:\Users\FML\anaconda3\envs\hnn_gui\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\FML\anaconda3\envs\hnn_gui\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\FML\anaconda3\envs\hnn_gui\Scripts\hnn-gui.exe\__main__.py", line 7, in <module>
  File "C:\Users\FML\anaconda3\envs\hnn_gui\lib\site-packages\hnn_core\gui\gui.py", line 1361, in launch
    from voila.app import main
ModuleNotFoundError: No module named 'voila'



I found it in a site (https://github.com/voila-dashboards/voila/archive/refs/tags/v0.3.6.zip) and use "pip install". The following error occured inside hnn_gui environment:

File "C:\Users\FML\anaconda3\envs\hnn_gui\lib\subprocess.py", line 373, in check_call
          raise CalledProcessError(retcode, cmd)
      subprocess.CalledProcessError: Command '['C:\\Users\\FML\\AppData\\Local\\Temp\\pip-build-env-v4qfubci\\overlay\\Scripts\\jlpm.EXE', 'install']' returned non-zero exit status 1.
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for voila
Failed to build voila
ERROR: Could not build wheels for voila, which is required to install pyproject.toml-based projects

So, I did the installation on the "base" environment, which runned fine.
However, I do not find voila module to install in hnn_gui env using the anaconda navigator.
Maybe it is only a question of what are the voila-related packages to look for and install.

Can you help me on this? It seems that it may be the only thing missing to get the gui working.

Best,

Sofia

**Comment by jasmainak** [comment, marked as answer]
Sofia,

You need to follow the instructions exactly as [provided here](https://jonescompneurolab.github.io/hnn-core/stable/index.html). These are the instructions we have tested ... any other variation may not work.

What is happening is that you are doing:

```sh
pip install hnn_core
```

and that does not install voila. If you try to install it on your own, you will end up installing the wrong version. You need to do:

```sh
pip install hnn_core[gui]
```

to ensure that it installs the right version of voila. So, let's try again in a fresh environment:

```sh
conda create -n hnn_gui2 python=3.9
```

then:

```sh
conda activate hnn_gui2
```

After this, no need to uninstall anything. The Neuron installation stays as it is. Just do:

```sh
pip install hnn_core[gui]
```

Finally to start the GUI, just do:

```sh
hnn-gui
```

Does this work? If not, what error do you get?

**Comment by srcfernandes** [comment]
Dear Jas,


It worked finally, thank you!

I was doing

pip install --upgrade https://api.github.com/repos/jonescompneurolab/hnn-core/zipball/master


and


pip install hnn_core[gui]


which was probably also not helping.


It still outputs some warnings when I open the hnn-gui, I hope this is not a problem:


[Voila] Using C:\Users\FML\AppData\Local\Temp to store connection files
[Voila] Storing connection files in C:\Users\FML\AppData\Local\Temp\voila_cabr3d8x.
[Voila] Serving static files from C:\Users\FML\anaconda3\envs\hnn_gui2\lib\site-packages\voila\static.
[Voila] Voilà is running at:
http://localhost:8866/
[Voila] WARNING | Notebook hnn_widget.ipynb is not trusted
[Voila] Kernel started: 281785cc-b7a7-4e65-87f3-bb7d50452e38
403 GET /voila/files/favicon.ico (::1): File not whitelisted
403 GET /voila/files/favicon.ico (::1) 2.00ms
C:\Users\FML\anaconda3\envs\hnn_gui2\lib\site-packages\jupyter_server\base\handlers.py:197: UserWarning: The Tornado web application does not have an 'authorizer' defined in its settings. In future releases of jupyter_server, this will be a required key for all subclasses of `JupyterHandler`. For an example, see the jupyter_server source code for how to add an authorizer to the tornado settings: https://github.com/jupyter-server/jupyter_server/blob/653740cbad7ce0c8a8752ce83e4d3c2c754b13cb/jupyter_server/serverapp.py#L234-L256
  warnings.warn(

Thank you so much! I was trying a lot of things and even starting to change the handlers.py file...


It would be nice to have these straighforward commands available to help others ones just starting as I am.


Many, many thanks for all your help and kindness!


Best,


Sofia





________________________________

**Comment by jasmainak** [comment]
Great! I think you can safely ignore the warnings for now. If there are others seeking help, can you redirect them to the discussions page: https://github.com/jonescompneurolab/hnn-core/discussions ? I converted this issue to a discussion. We are trying to separate developer stuff (issues) from user questions (discussions). Thanks!

**Comment by srcfernandes** [comment]
Yes, of course!


Will I get notifications to redirect them whenever necessary?



________________________________

**Comment by jasmainak** [comment]
Just share the URL :) That should be enough!
